src/modules/roc_netio/target_uv/roc_netio/ielement.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_netio/target_uv/roc_netio/ielement.h
+//! @brief close handler.
+
+#ifndef ROC_NETIO_IELEMENT_H_
+#define ROC_NETIO_IELEMENT_H_
+
+#include "roc_core/list_node.h"
+#include "roc_core/refcnt.h"
+
+#include "roc_packet/address.h"
+
+namespace roc {
+namespace netio {
+
+//! Network element interface.
+class IElement : public core::RefCnt<IElement>, public core::ListNode {
+public:
+    virtual ~IElement();
+
+    //! Asynchronous stop.
+    //! @remarks
+    //!  Should be called from the event loop thread.
+    virtual void stop() = 0;
+
+    //! Get bind address.
+    virtual const packet::Address& address() const = 0;
+
+private:
+    friend class core::RefCnt<IElement>;
+
+    virtual void destroy() = 0;
```

Comment:
It seems that IElement is more than just an interface: it's an interface plus referenc counting plus a node for list management.

Thus it probably makes sense to make two further steps:

* rename IElement to BasicElement or something like this;

* add a constructor BasicElement(allocator); make destroy() non-virtual; implement it by calling allocator_.destroy(this); remove destroy() implementation from child classes;

This will eliminate one virtual call (currently both destroy and destructor are virtual; after this change only destructor will be virtual) and arguably will make the code cleaner.

Not sure if this will be really better. Just an idea.

Reply:
We can also replace term "element" with "port". I see it like this:

```
class BasicPort ... {
public:
    virtual const packet::Address& address() const = 0;

    virtual bool open(packet::Address& bind_address) = 0;

    virtual void async_close() = 0;
};

class UDPSendingPort : public BasicPort { ... };
class UDPReceivingPort : public BasicPort { ... };
```

Reply:
Done